File: php/46-permutations.php
<?php
class Solution {
    private $paths = [];
    /**
     * @param Integer[] $nums
     * @return Integer[][]
     */
    function permute($nums) {
        $this->comb($nums);
        return $this->paths;
    }
    function comb($nums, $val = []) {
        $count = count($nums);
        if ($count == 1) {
            $this->paths[] = array_merge($val, $nums);
            return;
        }
        $nums = array_values($nums);
        for ($i = 0; $i < $count; $i++) {
            $temp = $nums;
            $tempNums = $val;
            unset($temp[$i]);
            $tempNums[] = $nums[$i];
            $this->comb($temp, $tempNums);
        }
    }
}

File: go/46-permutations.go
package main

var res [][]int = make([][]int, 0)

func permute(nums []int) [][]int {
	res = [][]int{}
	comb(nums, []int{})
	return res
}

func comb(nums []int, val []int) {
	if len(nums) == 1 {
		n := nums[0]
		r := append(val, n)
		copyVal := make([]int, len(r))
		copy(copyVal, r)
		res = append(res, copyVal)
		return
	}
	for k, v := range nums {
		sliceNums := make([]int, k)
		copy(sliceNums, nums[0:k])
		sliceNums = append(sliceNums, nums[k+1:]...)
		comb(sliceNums, append(val, v))
	}
}
